Implement the Illumina refinery

wvauclain opened the pull request:
## Methods

To determine which gene to map a given probe to, we select the genes in order of priority as follows:

- Pick the gene with the most appearances in Brainarray packages for Affymetrix platforms on the same species as the input Illumina platform
- If none of the associated gene IDs appear in any Brainarray platform, an NA is emitted
- If two or more of the associated gene IDs appear an equal number of times in
  Brainarray platforms, the gene ID with the lower number is selected to break the tie.

wvauclain committed:
Created a pipeline for picking out of duplicate gene IDs

wvauclain committed:
Add a readme

wvauclain committed:
Fix README formatting issue

wvauclain committed:
Add a section to the README about running the workflow

wvauclain committed:
Fix gene ID filtering to also recognize mouse and rat genes

jaclyn-taroni commented on 01_scraping_brainarray_genes/scrape_brainarray_genes.R:
```diff
@@ -0,0 +1,73 @@
+suppressPackageStartupMessages(library(lazyeval))
+suppressPackageStartupMessages(library(dplyr))
+
+ensg_url <- "http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/22.0.0/ensg.asp"
```
Hmm, should we make the BrainArray version a command line argument to make this easier to potentially update?

jaclyn-taroni commented on 01_scraping_brainarray_genes/scrape_brainarray_genes.R:
```diff
@@ -0,0 +1,73 @@
+
+    data.frame(mapped_gene_ids) %>%
+        dplyr::count(mapped_gene_ids) %>%
+        write.table(paste0("/out/01_scraping_brainarray_genes/", species, ".tsv"), row.names=FALSE, col.names=TRUE, quote=FALSE, sep="\t")
```
If you have `readr` installed already, `readr::write_tsv()` is generally more pleasant to work with!

kurtwheeler commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+out/*
+!out/02_selecting_gene_ids/
```
Missing newline here and in a few other files.

wvauclain committed:
Add a pre-commit config and fix trailing newlines

wvauclain committed:
Add license and citations

kurtwheeler commented on 02_selecting_gene_ids/select_gene_ids.py:
```diff
@@ -0,0 +1,107 @@
+  - Group the duplicate probes dataframe by probe ID
+  - For each group, pick the gene ID with the highest occurrence count. If there
+    are ties, then we want to pick the gene ID that is on the main Ensembl
+    assembly. If there are multiple genes on the main assembly, pick the one with
```
> If there are ties, then we want to pick the gene ID that is on the main Ensembl assembly

This isn't reflected in the section of the readme that outlines the priority.

wvauclain replied:
Good catch; this was our original decision for priority but this didn't end up actually breaking any ties and the implementation would not have been straightforward, so I removed that check. I will update this comment.

kurtwheeler commented on 02_selecting_gene_ids/select_gene_ids.py:
```diff
@@ -0,0 +1,107 @@
+
+        # if len(close_occurrences) == 1:
+        #     # No ties
+        #     return close_occurrences[0]
```
This code can be removed?

kurtwheeler commented on 02_selecting_gene_ids/select_gene_ids.py:
```diff
@@ -0,0 +1,107 @@
+        # # having at least half the # of occurrences as the max, because there
+        # # are some occurrences like [('ENSG00000143248', 29), ('ENSG00000232995', 3)]
+        # # which definitely are not ties, but others like
+        # # [('ENSG00000230876', 11), ('ENSG00000236854', 7)] which plausibly are
```
This seems to be doubly commented out.

kurtwheeler reviewed (approved):
This looks good to me. I think I understand what the R code is doing. I can't say I'd know it if there was a better way to do something though.

wvauclain committed:
Remove some unused commented-out code

wvauclain committed:
Update the header comment to reflect the new prioritization

jaclyn-taroni commented on README.md:
```diff
@@ -0,0 +1,57 @@
+
+To determine which gene to map a given probe to, we select the genes in order of priority as follows:
+
+- Pick the gene with the most appearances in Brainarray packages for Affymetrix platforms on the same species as the input Illumina platform
```
```suggestion
- Pick the gene with the most appearances in Brainarray packages for Affymetrix platforms of the same species as the input Illumina platform
```

jaclyn-taroni commented on README.md:
```diff
@@ -0,0 +1,57 @@
+- Mark Dunning, Andy Lynch and Matthew Eldridge (2021). illuminaMousev2.db: Illumina MouseWG6v2 annotation data (chip illuminaMousev2). R package version 1.26.0.
+- Mark Dunning, Andy Lynch and Matthew Eldridge (2021). illuminaRatv1.db: Illumina Ratv1 annotation data (chip illuminaRatv1). R package version 1.26.0.
+
+Other assets used in the computing of steo 01 come from the [Brainarray project](http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/genomic_curated_CDF.asp):
```
```suggestion
Other assets used in the computing of step 01 come from the [Brainarray project](http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/genomic_curated_CDF.asp):
```

jaclyn-taroni reviewed (approved):
LGTM - only some minor comments at this point

wvauclain commented:
@jaclyn-taroni I have one more question before I go ahead and merge this: is
> If none of the associated gene IDs appear in any Brainarray platform, an NA is emitted

a good behavior, or should we still select one of the gene IDs even when none of them appear in Brainarray?

jaclyn-taroni commented:
Hm, that seems like a case where we'd want to go to the main assembly but it sounds like that would have been quite the feat to implement? https://github.com/AlexsLemonade/illumina-refinery/pull/1#discussion_r675033383

wvauclain commented:
I can do some more digging, but from my initial impressions, the frontend says whether or not something is in the main assembly but I could not find anywhere on the API that indicates it.

jaclyn-taroni commented:
Gotcha - if that holds true, we should probably pick the lowest ID like we do to break ties.

wvauclain committed:
Allow trailing whitespace in our output TSV files

Pandas always outputs delimiters, even when they would be trailing
whitespace, so let's just allow it in our pre-commit hook.

wvauclain committed:
Add a command-line flag to set the brainarray version

wvauclain committed:
Correct README typos

wvauclain committed:
Use `readr::write_tsv`

wvauclain committed:
Fix typo in Makefile: actually use dummy file for step 01

wvauclain committed:
Add Ensembl checks to the tie-breaking code; avoid NAs

Previously, we were too overzealous with emitting NAs. Now, we don't
emit NAs if none of the genes appear in any Brainarray platform, and
instead we break ties by checking if any of the genes appear in the main
assembly on Ensembl. This method ends up breaking a lot of the ties.

wvauclain commented:
@jaclyn-taroni I dug through Ensembl's website's Perl code and figured out how they detect whether or not a gene is on the main assembly. It turns out that if we combine a couple of the API endpoints, we can approximate the same check, so I added that to the list of tiebreakers.

kurtwheeler commented on 02_selecting_gene_ids/select_gene_ids.py:
```diff
@@ -0,0 +1,140 @@
+            # If everything occurs 0 times, we will need to try another filter method
+            genes = [o[0] for o in occurrences]
+
+        # Now we have a tie. We need to gather more information on the genes in order ot progress.
```
```suggestion
        # Now we have a tie. We need to gather more information on the genes in order to progress.
```

wvauclain committed:
Correct comment typo

jaclyn-taroni commented:
Spot-checked a few of the human Ensembl gene IDs that were added in https://github.com/AlexsLemonade/illumina-refinery/pull/1/commits/00d567f4e4c344fe6dcd6d3426849d6a32a3ab6c on Ensembl's website and it looks as I would expect based on what you've included! 

Initially, it was less obvious to me (from looking at the website alone) that what we're retrieving for mouse and rat is equivalent. ([I think we'd expect only mouse and human to have this main (or primary) assembly, alternative distinction.](https://www.ensembl.org/info/genome/genebuild/haplotypes_patches.html
)). But it looks like the final "take the lower Ensembl ID" step is working as expected in the mouse and rat cases that initially gave me pause.

wvauclain commented:
@jaclyn-taroni I re-ran the mouse and rat platforms with some debugging print statements, and it looks like none of the mouse or rat genes come from alternate sequences.

The system does correctly detect that the tied gene IDs are both on the main platform, though, and applies the appropriate tiebreaker.
